File: web/types/alarm.ts
/**
 * 告警相关类型定义
 * 前端统一使用 camelCase 命名
 * 与后端 API 保持一致
 */

// 报警级别
export type AlarmLevel = 'Info' | 'Warning' | 'Error' | 'Critical'

// 报警状态
export type AlarmStatus = 'Active' | 'Acknowledged' | 'Resolved' | 'Suppressed'

// 报警类型
export type AlarmType = 'DeviceOffline' | 'DeviceError' | 'PropertyThreshold' | 'PropertyAnomaly' | 'CommandFailed' | string

// 解决方式
export type ResolutionType = 'Fixed' | 'FalseAlarm' | 'Ignored' | 'AutoResolved'

// 规则类型
export type RuleType = 'threshold' | 'range' | 'change' | 'duration' | 'composite'

// 比较运算符
export type ComparisonOperator = 'greater_than' | 'less_than' | 'greater_than_or_equal' | 'less_than_or_equal' | 'equal' | 'not_equal'

// 变化类型
export type ChangeType = 'increase' | 'decrease' | 'any'

// 逻辑运算符
export type LogicalOperator = 'and' | 'or' | 'not'

// 报警条件 - 使用内部标签格式
export type AlarmCondition = 
  | { type: 'threshold'; operator: ComparisonOperator; value: number }
  | { type: 'range'; min?: number; max?: number; inclusive: boolean }
  | { type: 'change'; changeType: ChangeType; threshold: number; timeWindow: number }
  | { type: 'duration'; condition: AlarmCondition; duration: number }
  | { type: 'composite'; operator: LogicalOperator; conditions: AlarmCondition[] }

// 通知渠道类型
export type NotificationChannelType = 'Email' | 'Sms' | 'Webhook' | 'Sse'

// 通知配置
export interface NotificationConfig {
  enabled: boolean
  channels: NotificationChannelType[]
  recipients: string[]
  suppressDuration?: number
  repeatInterval?: number
}

// 确认信息
export interface Acknowledgement {
  acknowledgedBy: string
  acknowledgedAt: string
  note?: string
}

// 解决信息
export interface Resolution {
  resolvedBy: string
  resolvedAt: string
  note?: string
  resolutionType: ResolutionType
}

// 报警实例
export interface Alarm {
  id: string
  deviceId: string
  deviceName?: string
  propertyId?: string
  propertyName?: string
  ruleId?: string
  ruleName?: string
  alarmType: string
  alarmLevel: string
  message: string
  alarmValue?: string
  thresholdValue?: string
  alarmTime: string
  status: string
  isAcknowledged: boolean
  acknowledgedBy?: string
  acknowledgedAt?: string
  acknowledgedNote?: string
  isResolved: boolean
  resolvedBy?: string
  resolvedAt?: string
  resolvedNote?: string
  createdAt: string
}

// 报警规则
export interface AlarmRule {
  id: string
  name: string
  description?: string
  deviceId?: string
  propertyId?: string
  ruleType: string
  condition: AlarmCondition
  alarmLevel: string
  isEnabled: boolean
  notificationConfig: NotificationConfig
  createdAt: string
  updatedAt: string
}

// 报警统计
export interface AlarmStatistics {
  totalCount: number
  activeCount: number
  acknowledgedCount: number
  resolvedCount: number
}

// 查询参数
export interface AlarmQueryParams {
  deviceIds?: string[]
  levels?: string[]
  statuses?: string[]
  startTime?: string
  endTime?: string
  page?: number
  pageSize?: number
}

export interface StatisticsQueryParams {
  startTime?: string
  endTime?: string
}

// 创建规则请求
export interface CreateAlarmRuleRequest {
  name: string
  description?: string
  deviceId?: string
  propertyId?: string
  ruleType: RuleType
  condition: AlarmCondition
  alarmLevel: AlarmLevel
  notificationConfig: NotificationConfig
}

// 更新规则请求
export interface UpdateAlarmRuleRequest {
  name?: string
  description?: string
  condition?: AlarmCondition
  alarmLevel?: AlarmLevel
  notificationConfig?: NotificationConfig
}

// 确认请求
export interface AcknowledgeRequest {
  note?: string
}

// 解决请求
export interface ResolveRequest {
  resolutionType: ResolutionType
  note?: string
}

// 批量确认请求
export interface BatchAcknowledgeRequest {
  alarmIds: string[]
}

// 批量解决请求
export interface BatchResolveRequest {
  alarmIds: string[]
  resolutionType: ResolutionType
}

// 批量操作结果
export interface BatchOperationResult {
  successCount: number
  totalCount: number
}

File: crates/tinyiothub-core/src/models/alarm.rs
use serde::{Deserialize, Serialize};

/// Alarm DTO
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AlarmDto {
    pub id: String,
    pub device_id: String,
    pub device_name: Option<String>,
    pub property_id: Option<String>,
    pub property_name: Option<String>,
    pub rule_id: Option<String>,
    pub rule_name: Option<String>,
    pub alarm_type: String,
    pub alarm_level: String,
    pub message: String,
    pub alarm_value: Option<String>,
    pub threshold_value: Option<String>,
    pub alarm_time: String,
    pub status: String,
    pub is_acknowledged: bool,
    pub acknowledged_by: Option<String>,
    pub acknowledged_at: Option<String>,
    pub acknowledged_note: Option<String>,
    pub is_resolved: bool,
    pub resolved_by: Option<String>,
    pub resolved_at: Option<String>,
    pub resolved_note: Option<String>,
    pub created_at: String,
}

/// Alarm rule DTO
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AlarmRuleDto {
    pub id: String,
    pub name: String,
    pub description: Option<String>,
    pub device_id: Option<String>,
    pub device_name: Option<String>,
    pub property_id: Option<String>,
    pub property_name: Option<String>,
    pub rule_type: String,
    pub condition: serde_json::Value,
    pub alarm_level: String,
    pub is_enabled: bool,
    pub notification_config: serde_json::Value,
    pub created_at: String,
    pub updated_at: String,
}

/// Alarm statistics DTO
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AlarmStatisticsDto {
    pub total_count: u64,
    pub active_count: u64,
    pub acknowledged_count: u64,
    pub resolved_count: u64,
}
